### PR TITLE
Add spread field completion support

### LINKED ISSUE #25383: Add completion support for the spread-field
**Description:**
Add the completion support for the spread field, where the LS should suggest the possible mapping variables.

### PR DESCRIPTION
**Pull request by malinthar**
## Purpose
$subject

Fixes #25383

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java [hunk 1]
```diff
@@ -187,7 +190,77 @@ protected boolean hasReadonlyKW(Node evalNodeAtCursor) {
...
+                    }
+                    return true;
+                })).collect(Collectors.toList());
+
+        return getSpreadFieldCompletionItemList(visibleSymbols, context);
+    }
+
+    private List<LSCompletionItem> getSpreadFieldCompletionItems(BallerinaCompletionContext context,
+                                                                 List<RecordFieldSymbol> validFields,
+                                                                 RecordTypeSymbol recordTypeSymbol) {
```
Comment (gayalkuruppu) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java:224]: Is this parameter `recordTypeSymbol` necessary?

**Review by gayalkuruppu — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java [hunk 2]
```diff
@@ -187,7 +190,77 @@ protected boolean hasReadonlyKW(Node evalNodeAtCursor) {
...
+        List<Symbol> visibleSymbols = context.visibleSymbols(context.getCursorPosition()).stream()
+                .filter(this.getVariableFilter().and(symbol -> {
```
Comment (TharushiJay) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java:226]: Are we only providing spread field completion support for variables? 
What about the following?
```
type Employee record {|
    string name;
|};

public function main() {
    Employee e = {...getEmployee()};
}

function getEmployee() returns Employee {
    return {name:""};
}
```

Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java:226]: +1. I have missed this

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/builder/SpreadFieldCompletionItemBuilder.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+public class SpreadFieldCompletionItemBuilder {
+
+    /**
+     * Build the constant {@link CompletionItem}.
+     *
+     * @param symbol  {@link Symbol}
+     * @param context Completion context
+     * @return {@link CompletionItem} generated completion item
+     */
+    public static CompletionItem build(Symbol symbol, String typeName, BallerinaCompletionContext context) {
```
[the remark this answers is not in the document]
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/builder/SpreadFieldCompletionItemBuilder.java:39]: No it's not. I will remove it

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java (hunk 1 above)
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java:224]: No it's not. Will remove it

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java (hunk 2 above)
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java]: Done

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/builder/SpreadFieldCompletionItemBuilder.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+     *
+     * @param symbol   {@link FunctionSymbol}
+     * @param typeName type name of the {@link Symbol}
+     * @param context  completion context
+     * @return {@link CompletionItem} generated completion item
+     */
+    public static CompletionItem build(FunctionSymbol symbol, String typeName, BallerinaCompletionContext context) {
+        String symbolName = symbol.getName().orElseThrow();
+        Pair<String, String> functionInvocationSignature =
+                FunctionCompletionItemBuilder.getFunctionInvocationSignature(symbol, symbolName, context);
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/builder/SpreadFieldCompletionItemBuilder.java:58]: Avoid using `FunctionCompletionItemBuilder.getFunctionInvocationSignature`. We are not supposed to use the method as a utility method. If there is a need, we have to move it properly to a utility.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/builder/SpreadFieldCompletionItemBuilder.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                                        String typeName) {
+        CompletionItem completionItem = new CompletionItem();
+        completionItem.setLabel(label);
+        completionItem.setKind(kind);
+        completionItem.setDetail(typeName);
+        completionItem.setFilterText(symbolName);
+        completionItem.setInsertText(insertText);
+        return completionItem;
+    }
+
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/builder/SpreadFieldCompletionItemBuilder.java:84]: Extra space

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java [hunk 6]
```diff
@@ -187,7 +194,95 @@ protected boolean hasReadonlyKW(Node evalNodeAtCursor) {
...
+                        rawType = CommonUtil.getRawType(returnTypeSymbol.get());
+                    } else {
+                        rawType = CommonUtil.getRawType(typeDescriptor.get());
+                    }
+                    if (rawType.typeKind() != TypeDescKind.RECORD) {
+                        return false;
+                    }
+                    Map<String, RecordFieldSymbol> recordFields = ((RecordTypeSymbol) rawType).fieldDescriptors();
+                    for (RecordFieldSymbol fieldSymbol : recordFields.values()) {
+                        if (fieldSymbol.getName().isEmpty() || validFields.stream()
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java:249]: Try to move the filter-based check to private util for better readability

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java [hunk 7]
```diff
@@ -187,7 +194,95 @@ protected boolean hasReadonlyKW(Node evalNodeAtCursor) {
...
+                                .filter(validField -> validField.getName().isPresent()
+                                        && validField.getName().get().equals(fieldSymbol.getName().get())
+                                        && fieldSymbol.typeDescriptor()
+                                        .subtypeOf(validField.typeDescriptor())).findAny().isEmpty()) {
+                            return false;
+                        }
+                    }
+                    return true;
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java:257]: We have multiple returns which is okay, but given the logic it is bit complex to understand. Check whether if we can simplify this

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/builder/SpreadFieldCompletionItemBuilder.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+import io.ballerina.compiler.api.symbols.Symbol;
+import org.ballerinalang.langserver.commons.BallerinaCompletionContext;
+import org.eclipse.lsp4j.CompletionItem;
+import org.eclipse.lsp4j.CompletionItemKind;
+
+/**
+ * This class is being used to build spread field completion item.
+ *
+ * @since 2.0.0
```
[the remark this answers is not in the document]
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/builder/SpreadFieldCompletionItemBuilder.java]: Done

**Review by mohanvive — DISMISSED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java [hunk 9]
```diff
@@ -187,7 +194,99 @@ protected boolean hasReadonlyKW(Node evalNodeAtCursor) {
...
+        rawType = CommonUtil.getRawType(typeDescriptor.get());
+        if (rawType.typeKind() != TypeDescKind.RECORD) {
+            return false;
+        }
+        Map<String, RecordFieldSymbol> recordFields = ((RecordTypeSymbol) rawType).fieldDescriptors();
+        for (RecordFieldSymbol fieldSymbol : recordFields.values()) {
+            if (!isAMember(fieldSymbol, fields)) {
+                return false;
+            }
+        }
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java:257]: Shall we use the stream API here? Maybe we can go with `anyMatch`? So we can have single return statement.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java [hunk 10]
```diff
@@ -187,7 +194,99 @@ protected boolean hasReadonlyKW(Node evalNodeAtCursor) {
...
+                typeDescriptor = ((FunctionTypeSymbol) typeDescriptor.get()).returnTypeDescriptor();
+            }
+            String typeName = (typeDescriptor.isEmpty() || typeDescriptor.get().typeKind() == null) ? "" :
+                    CommonUtil.getModifiedTypeName(ctx, typeDescriptor.get());
+            CompletionItem cItem;
+            if (symbol.kind() == FUNCTION) {
+                cItem = SpreadFieldCompletionItemBuilder.build((FunctionSymbol) symbol, typeName, ctx);
+            } else {
+                cItem = SpreadFieldCompletionItemBuilder.build(symbol, typeName);
+            }
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java:286]: What if we do this check inside the `SpreadFieldCompletionItemBuilder` to  maintain SRP?

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java [hunk 11]
```diff
@@ -1910,4 +1911,65 @@ public static String getModifiedUri(WorkspaceManager workspaceManager, String ur
...
+    public static List<String> getFuncArguments(FunctionSymbol symbol, BallerinaCompletionContext ctx) {
+        List<String> args = new ArrayList<>();
+        boolean skipFirstParam = skipFirstParam(ctx, symbol);
+        FunctionTypeSymbol functionTypeDesc = symbol.typeDescriptor();
+        Optional<ParameterSymbol> restParam = functionTypeDesc.restParam();
+        List<ParameterSymbol> parameterDefs = new ArrayList<>();
+
+        if (functionTypeDesc.params().isPresent()) {
+            List<ParameterSymbol> params = functionTypeDesc.params().get();
+            if (params.size() > 1 && skipFirstParam) {
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java:1931]: What if we have only one parameter and still we have to skip that also. For instance consider the `length()` function. We need to handle that also.

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java [hunk 12]
```diff
@@ -1910,4 +1911,65 @@ public static String getModifiedUri(WorkspaceManager workspaceManager, String ur
...
     }
+
+    /**
+     * Get the list of function arguments from the invokable symbol.
+     *
+     * @param symbol Invokable symbol to extract the arguments
+     * @param ctx    Lang Server Operation context
+     * @return {@link List} List of arguments
+     */
+    public static List<String> getFuncArguments(FunctionSymbol symbol, BallerinaCompletionContext ctx) {
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java:1922]: What if we make this function return `List<ParameterSymbol>` and write another util to get a list of modified type names?

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java [hunk 13]
```diff
@@ -187,10 +192,93 @@ protected List<LSCompletionItem> getFieldCompletionItems(BallerinaCompletionCont
...
+
+    private boolean isAMember(RecordFieldSymbol field, List<RecordFieldSymbol> fields) {
+        return field.getName().isPresent() && fields.stream().anyMatch(validField ->
+                validField.getName().isPresent() && validField.getName().get().equals(field.getName().get())
+                        && field.typeDescriptor().subtypeOf(validField.typeDescriptor()));
+    }
+
+    private List<LSCompletionItem> getSpreadFieldCompletionItemList(List<Symbol> visibleSymbols,
+                                                                    BallerinaCompletionContext ctx) {
+        List<Symbol> processedSymbols = new ArrayList<>();
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java:262]: Can we use a `Set` instead?

**Review by IMS94 — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java (hunk 12 above)
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java:1922]: +1

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java (hunk 13 above)
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java:262]: +1

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java [hunk 14]
```diff
@@ -1998,4 +2000,94 @@ public static Optional<Node> getMatchingNode(Node currentNode, Predicate<Node> p
...
+
+        if (functionTypeDesc.params().isPresent()) {
+            List<ParameterSymbol> params = functionTypeDesc.params().get();
+            if (skipFirstParam) {
+                if (params.size() > 1) {
+                    parameterDefs.addAll(params.subList(1, params.size()));
+                }
+            } else {
+                parameterDefs.addAll(params);
+            }
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java:2071]: `parameterDefs.addAll(params.subList(skipFirstParam ? 1 : 0, params.size()))`

What about something like the above? Would that satisfy the requirement?

**Review by nadeeshaan — DISMISSED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java (hunk 14 above)
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java:2071]: The suggested code snippet won't work as we have to check for the size of the params before slicing. Adding that condition into a conditional expression will further complicate the logic IMO.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java [hunk 15]
```diff
@@ -1998,4 +2000,94 @@ public static Optional<Node> getMatchingNode(Node currentNode, Predicate<Node> p
...
+    }
+
+    /**
+     * Get the parameter type and name pair given the paramter symbol.
+     *
+     * @param param parameter symbol.
+     * @param ctx   Lang Server Operation context
+     * @return {@link Optional<String>} Type and name pair.
+     */
+    public static Optional<String> getParamTypeAndNamePair(ParameterSymbol param, BallerinaCompletionContext ctx) {
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java:2031]: Does the name of this method still valid?

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java [hunk 16]
```diff
@@ -1998,4 +2000,94 @@ public static Optional<Node> getMatchingNode(Node currentNode, Predicate<Node> p
...
+    }
+
+    /**
+     * Get the list of function parameters from the invokable symbol.
+     *
+     * @param symbol Invokable symbol to extract the parameters
+     * @param ctx    Lang Server Operation context
+     * @return Pair of list of parameter symbols and rest parameter symbol.
+     */
+    public static Pair<List<ParameterSymbol>, Optional<ParameterSymbol>> getFunctionParameters(
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java:2056]: I remember we agreed to return a list of `ParameterSymbol` instead and determine if a parameter is a `REST` param using parameter kind.

**Review by IMS94 — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java (hunk 15 above)
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java:2031]: This need to be fixed

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java (hunk 16 above)
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java:2056]: Will fix this. +1

**Review by IMS94 — APPROVED** [review]

**Review by IMS94 — CHANGES_REQUESTED** [review]

**Review by IMS94 — APPROVED** [review]

**Pull request merged**